mne/filter.py

```diff
@@ -376,49 +376,86 @@ def _filter(x, Fs, freq, gain, filter_length='10s', picks=None, n_jobs=1,
-    filtfilt = get_filtfilt()
+    padlen = min(iir_params['padlen'], len(x))
     n_jobs = check_n_jobs(n_jobs)
     x, orig_shape, picks = _prep_for_filtering(x, copy, picks)
-    _check_coefficients(b, a)
+    if 'sos' in iir_params:
+        sosfiltfilt = get_sosfiltfilt()
+        fun = partial(sosfiltfilt, sos=iir_params['sos'], padlen=padlen)
+        _check_coefficients(iir_params['sos'])
+    else:
+        filtfilt = get_filtfilt()
+        fun = partial(filtfilt, b=iir_params['b'], a=iir_params['a'],
+                      padlen=padlen)
+        _check_coefficients((iir_params['b'], iir_params['a']))
     if n_jobs == 1:
         for p in picks:
-            x[p] = filtfilt(b, a, x[p], padlen=padlen)
+            x[p] = fun(x=x[p])
     else:
-        parallel, p_fun, _ = parallel_func(filtfilt, n_jobs)
-        data_new = parallel(p_fun(b, a, x[p], padlen=padlen)
-                            for p in picks)
+        parallel, p_fun, _ = parallel_func(fun, n_jobs)
+        data_new = parallel(p_fun(x=x[p]) for p in picks)
         for pp, p in enumerate(picks):
             x[p] = data_new[pp]
     x.shape = orig_shape
     return x
 
 
-def _estimate_ringing_samples(b, a):
-    """Helper function for determining IIR padding"""
-    # XXX Need to extend this to more than 1000 samples for long IIR filters!
-    from scipy.signal import lfilter
-    x = np.zeros(1000)
+def _estimate_ringing_samples(system):
+    """Helper function for determining filter ringing (up to 100000 samples)"""
+    from scipy import signal
+    if isinstance(system, tuple):  # TF
+        kind = 'ba'
+        b, a = system
+        zi = [0.] * (len(a) - 1)
+    else:
+        kind = 'sos'
+        sos = system
+        zi = [[0.] * 2] * len(sos)
+    n_per_chunk = 1000
+    n_chunks_max = 100
+    x = np.zeros(n_per_chunk)
     x[0] = 1
-    h = lfilter(b, a, x)
-    return np.where(np.abs(h) > 0.001 * np.max(np.abs(h)))[0][-1]
+    last_good = 1000
+    thresh_val = 0
+    for ii in range(n_chunks_max):
+        if kind == 'ba':
+            h, zi = signal.lfilter(b, a, x, zi=zi)
+        else:
+            h, zi = signal.sosfilt(sos, x, zi=zi)
+        x[0] = 0  # for subsequent iterations we want zero input
+        h = np.abs(h)
+        thresh_val = max(0.001 * np.max(h), thresh_val)
+        idx = np.where(np.abs(h) > thresh_val)[0]
+        if len(idx) > 0:
+            last_good = idx[-1]
+        else:  # this iteration had no sufficiently lange values
+            idx = (ii - 1) * 1000 + last_good
+            break
+    else:
+        warn('Could not properly estimate ringing for the filter')
+        idx = n_per_chunk * n_chunks_max
+    return idx
 
 
-def construct_iir_filter(iir_params=dict(b=[1, 0], a=[1, 0], padlen=0),
-                         f_pass=None, f_stop=None, sfreq=None, btype=None,
-                         return_copy=True):
+def construct_iir_filter(iir_params, f_pass=None, f_stop=None, sfreq=None,
```

Comment:
Really it shouldn't have a default. The `iir_params` that gets passed in just needs not to have an `'output'` key and the warning will appear.